flang/lib/Lower/Allocatable.cpp

```diff
@@ -42,12 +42,76 @@ static llvm::cl::opt<bool> useDescForMutableBox(
     llvm::cl::desc("Always use descriptors for POINTER and ALLOCATABLE"),
     llvm::cl::init(false));
 
+//===----------------------------------------------------------------------===//
+// Error management
+//===----------------------------------------------------------------------===//
+
+namespace {
+// Manage STAT and ERRMSG specifier information across a sequence of runtime
+// calls for an ALLOCATE/DEALLOCATE stmt.
+struct ErrorManager {
+  void init(Fortran::lower::AbstractConverter &converter, mlir::Location loc,
+            const Fortran::lower::SomeExpr *statExpr,
+            const Fortran::lower::SomeExpr *errMsgExpr) {
+    Fortran::lower::StatementContext stmtCtx;
+    auto &builder = converter.getFirOpBuilder();
+    hasStat = builder.createBool(loc, statExpr != nullptr);
+    statAddr = statExpr
+                   ? fir::getBase(converter.genExprAddr(statExpr, stmtCtx, loc))
+                   : mlir::Value{};
+    if (statExpr && errMsgExpr) {
+      auto errMsgDesc = builder.createBox(
+          loc, converter.genExprAddr(errMsgExpr, stmtCtx, loc));
+      errMsgAddr = builder.createTemporary(loc, errMsgDesc.getType());
+      builder.create<fir::StoreOp>(loc, errMsgDesc, errMsgAddr);
```

Comment:
When you update the `const Descriptor *` in the runtime, this will need to be revisited.